segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/WriteOperation.java

```diff
@@ -225,7 +241,7 @@ private CompletableFuture<Void> writeData(MetadataTransaction txn) {
                                 // Write data to last chunk.
                                 return writeToChunk(txn,
                                         segmentMetadata,
-                                        data,
+                                        inputStream,
```

Comment:
Here inputStream is being passed with the same content as inputData which is doing the validation. So rather than setting it before the call here, it could be passed as a parameter and then there would be no need for the `inputData` member variable.

Reply:
Here key issue is that original InputStream passed in is NOT resettable. You can not read it and go back and read it gain for validation.
So instead what we are doing is reading it whole in an array and storing it in  `inputData`, then we wrap this new array in ByteArrayInputStream and pass it on.

The validation code then just uses already captured byte array for comparing against data read from LTS. During validation comparing arrays is easy and faster.

Note that just for validation we need to allocate arrays which has perf penalty. We do not want to do this during normal operation.

Also we can't change upper layers or lower layers - change has to be pretty local and self contained.

Reply:
updated as discussed.